upb/json/parser.rl

```diff
@@ -1325,27 +1457,103 @@ static void end_object(upb_json_parser *p) {
 
   value =
     number
-      >{ start_number(parser, p); }
-      %{ CHECK_RETURN_TOP(end_number(parser, p)); }
+      >{
+         if (is_top_level(parser)) {
+           CHECK_RETURN_TOP(is_number_wrapper_object(parser));
+           start_wrapper_object(parser);
+         } else if (does_number_wrapper_start(parser)) {
+           CHECK_RETURN_TOP(start_subobject(parser));
+           start_wrapper_object(parser);
+         }
+         start_number(parser, p);
+       }
+      %{
+         CHECK_RETURN_TOP(end_number(parser, p));
+         if (does_number_wrapper_end(parser)) {
+           end_wrapper_object(parser);
+           if (!is_top_level(parser)) {
+             end_subobject(parser);
+           }
+         }
+       }
     | string
-      >{ CHECK_RETURN_TOP(start_stringval(parser)); }
-      @{ CHECK_RETURN_TOP(end_stringval(parser)); }
+      >{
+         if (is_top_level(parser)) {
+           CHECK_RETURN_TOP(is_string_wrapper_object(parser));
+           start_wrapper_object(parser);
+         } else if (does_string_wrapper_start(parser)) {
+           CHECK_RETURN_TOP(start_subobject(parser));
+           start_wrapper_object(parser);
+         }
+         CHECK_RETURN_TOP(start_stringval(parser));
+       }
+      @{
+         CHECK_RETURN_TOP(end_stringval(parser));
+         if (does_string_wrapper_end(parser)) {
+           end_wrapper_object(parser);
+           if (!is_top_level(parser)) {
+             end_subobject(parser);
+           }
+         }
+       }
     | "true"
-      %{ CHECK_RETURN_TOP(parser_putbool(parser, true)); }
+      >{
+         if (is_top_level(parser)) {
```

Comment:
Can we wait until the end of "true" to do all this work? It seems like all of the start/end code for boolean could go in parser_putbool().

Likewise for "false" below. We only care that we say true/false. We don't need to do anything at the beginning of the word.